### ISSUE TITLE
Method overloads and type specificity are influenced by order of definition

### ISSUE THREAD

**Issue by jgaskins** [crystal-lang/crystal#14916]
If I define a type hierarchy before a set of method overloads that receive them, the correct method is invoked:

```crystal
abstract struct Event
end

struct MessageEvent < Event
end

def handle(event : Event)
  pp unhandled_event: event, type: typeof(event)
end

def handle(event : MessageEvent)
  pp message_event: event
end

handle MessageEvent.new
# {message_event: MessageEvent()}
```

The `message_event` key in the output indicates it's calling the `MessageEvent` overload. If the _methods_ are defined first, that changes:

```crystal
def handle(event : Event)
  pp unhandled_event: event, type: typeof(event)
end

def handle(event : MessageEvent)
  pp message_event: event
end

abstract struct Event
end

struct MessageEvent < Event
end

handle MessageEvent.new
# {unhandled_event: MessageEvent(), type: MessageEvent}
```

This time it calls the `Event` overload, denoted by the `unhandled_event` key in the output, even though the type of the object at compile time is `MessageEvent`, not `Event`.

But if we reorder the methods in descending order of specificity, it calls the correct one.

```crystal
def handle(event : MessageEvent)
  pp message_event: event
end

def handle(event : Event)
  pp unhandled_event: event, type: typeof(event)
end

abstract struct Event
end

struct MessageEvent < Event
end

handle MessageEvent.new
# {message_event: MessageEvent()}
```

I've seen this happen a lot over the years, but I could never figure out why it happened sometimes but not others. I finally understand it, but every time it's happened, it's surprising and I have to spend a lot of time debugging it.

**Comment by HertzDevil** [comment]
Is this resolved by #11840?

**Comment by devnote-dev** [comment]
I haven't looked too deeply into this issue but I do wonder why it would be something that needs "fixing". For as long as I can remember this has always been expected behaviour and makes logical sense; I've seen this pattern used in quite a few shards and even use it in a lot of my own.

**Comment by jgaskins** [comment]
@HertzDevil It is! I had to merge the `master` branch into yours in order to compile the compiler to test, but it does give the expected output! Looks like this issue is a duplicate of #7579.

---

@devnote-dev While it may be expected and logical to you, that hasn't been the case for me. Most of the time, defining methods and types in arbitrary order just works. I only discovered what was causing this while trying to write a minimal reproduction of it to file this issue, despite having run into this problem for years. Now that I know what the problem is, it does seem logical, but until that point it's just been confusing and discouraging. In the past, I've always just ended up giving up and moved the superclass "catch-all" overload to the end after a few hours of screwing around with it.

If what I'm doing is unexpected or unsupported, ideally it would fail to compile (as it does when using a macro, including a mixin, or inheriting from a superclass before any of them are defined) rather than silently calling the wrong method.

It's also notable that I've only ever run into this when I try to write event handlers as methods with different event types accepted as overloads. I'm sure others have run into it in different scenarios, but for me it's always things like Slack events, GitHub webhooks, etc.
